### ISSUE TITLE
Sessions

### ISSUE THREAD

**Issue by hinell** [equalsraf/neovim-qt#1026]
Anyway we can configure neovim-qt to to not close when the system is shut down?

**Comment by jgehrig** [comment]
I don't have any plans to implement this feature soon...

However, it looks like you've put some thought and research into this; feel free to open a Pull Request!

If you can provide a clean and simple implementation, I will gladly provide guidance, review code, and merge it.

My potential hesitance/concerns:
 - Adding additional state/features make testing harder.h
 - We would need a mechanism to restore the `nvim` state too. I'm not sure if this would be simple or complicated?
 - If this requires X11/Linux specific code, I will become more hesitant to merge. Platform agnostic code is preferred.

**Comment by hinell** [comment]
>feel free to open a Pull Request!

@jgehrig Hi, thanks! I will consider to do this at some point for sure. We have had this feature in QT apps running under KDE DE for many years. It's been default X11-builtin for ages, take a look: [X session manager](https://en.wikipedia.org/wiki/X_session_manager). KDE folks even have implemented X.org's session server namely [`ksmserver`](https://www.linuxtopia.org/online_books/linux_desktop_guides/kde_desktop_user_guide/ksmserver.html). It's very useful feature as it makes devs more productive. When system is started up I just continue where I left.

>We would need a mechanism to restore the nvim state too

I think this could be as simple as a few commands restoring opened files. It doesn't require anything sofisticated.

>If this requires X11/Linux specific code

Qt provides a guideline on sessions implementation and they have this feature built-in into their framework. It doesn't provide it for Windows systems though because Windows doesn't have session manager like X11 or KDE do so it's going to be excluded anyway.

**Comment by hinell** [comment]
@equalsraf I agree, the native [`mksession`](https://neovim.io/doc/user/starting.html#views-sessions) or any other command may be utilized by users via `init.lua` configuration to manage session. This command saves `Session.vim` file.

>... Instead of trying to determine which context should be saved (or not)

Yeah, nvim's native `autocmd`s can also be used as hook (listener) into nvim session lifecycle from plugins.

#### Plugins list to manage sessions

A few weeks ago I've found there are plugins to manage sessions for neovim, some of them hook into neovim events and save sessions upon exit (i.e. olimorris/persisted), checkout:
* https://github.com/tpope/vim-obsession
* https://github.com/olimorris/persisted.nvim
* https://github.com/Shatur/neovim-session-manager

**Comment by equalsraf** [comment]
An autocmd would work for *saveStateStateRequest*. But not for *commitDataRequest* because it may need to return data (cancelation). But maybe it is best to skip implementing commitDataRequest for starters.


For saveStateStateRequest

```c++
m_nvim->api0()->vim_command("doautocmd User GuiSessionSaveRequest");
```

and another for restoring the session startup `GuiSessionRestore`.


Also need a way to expose the session id, like a global var `g:GuiSessionId`. This can be set at startup from `QGuiApplication::sessionId()`.
